refactor: dropping recipient from `NoteDao` and `UniqueNote`

benesjan opened the pull request:
Fixes F-92

The issue description describes the motivation for this well:

> NoteDao contains a recipient value. This is no longer used as there is no filtering by recipient. It is also incorrect as notes don't technically have a recipient, they're just data. It is the note message that has a recipient.
> 
> As recipient gets removed from the object, we'll no longer be able to use it to determine the note's scope. We should therefore ask the caller to specify the scope, which they can just do with the recipient field they used when creating the NoteDao (the message recipient).

benesjan commented:
* **#18586** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18586?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18586?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.test.ts:
```diff
@@ -657,11 +657,10 @@ describe('PXEOracleInterface', () => {
 
       // Verify note was stored
-      const notes = await noteDataProvider.getNotes({ contractAddress });
+      const notes = await noteDataProvider.getNotes({ contractAddress, scopes: [recipient.address] });
```
Now instead of manually filtering all notes we just request notes for the relevant scope.

benesjan commented on yarn-project/stdlib/src/note/unique_note.ts:
As I am looking at the code I think it doesn't make sense for `UniqueNote` to exist anymore. I would drop it and replace its use with `NoteDao`. Maybe I could also rename `NoteDao` to something sensible while doing that.

I think it make sense to have the term "note" term reserved for this struct or its serialized form:

```rust
#[note]
pub struct MyNote {
    value: u128,
}
```

so `Note` is not an option for `NoteDao`.

`NoteDao` is the note along with a ton of other data.

This is what AI came up with:

<img width="1123" height="531" alt="Image" src="https://github.com/user-attachments/assets/2c0ce5f2-1801-4cc0-94be-6ddca6ee9613" />

I think `CommittedNote` is quite nice. Would you agree?

nventuro replied:
`UniqueNote` is indeed bad, but the correct name is `SettledNote`, not `Committed`.  This comes from the kernels, see e.g. `NoteStageEnum` in `note_metadata.nr`.

This is incomplete however, it seems to be missing the randomness. Also, the tx hash is not part of note. We don't need to fix this right now though.

benesjan replied:
Dropped UniqueNote in https://github.com/AztecProtocol/aztec-packages/pull/18593

benesjan commented on yarn-project/pxe/src/pxe.ts:
```diff
@@ -679,19 +679,9 @@ export class PXE {
 
     const noteDaos = await this.noteDataProvider.getNotes(filter);
 
-    const uniqueNotes = noteDaos.map(async dao => {
-      const completeAddresses = await this.addressDataProvider.getCompleteAddresses();
-      const completeAddressIndex = completeAddresses.findIndex(completeAddress =>
-        completeAddress.address.equals(dao.recipient),
-      );
-      const completeAddress = completeAddresses[completeAddressIndex];
-      if (completeAddress === undefined) {
-        throw new Error(`Cannot find complete address for recipient ${dao.recipient.toString()}`);
-      }
-      const recipient = completeAddress.address;
-      return new UniqueNote(dao.note, recipient, dao.contractAddress, dao.storageSlot, dao.txHash, dao.noteNonce);
```
This was truly bizarre.

nventuro replied:
Yeah, this is quite insane. I looked at the git blame and it is also hard to find an older version of this where it makes sense.

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -194,10 +194,10 @@ export class NoteDataProvider {
 
         if (scopes.length === 0) {
-          scopes = [dao.recipient.toString()];
+          throw new Error(`No scopes found for nullified note with index ${noteIndex}`);
```
It should not ever happen that this would get throw because the only way a note can get into the provider is via addNotes and then when applying nullifiers the scope should get moved over.

nventuro replied:
This kind of thing should go in comments - left a suggestion

benesjan commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -359,7 +359,12 @@ export class NoteDataProvider {
+
+        const noteScopes = await toArray(this.#notesToScope.getValuesAsync(noteIndex));
+        if (noteScopes.length === 0) {
+          throw new Error('Note scopes are missing in applyNullifiers');
```
It should not ever happen that this would get thrown because the only way a note can get into the provider is via addNotes and that requires a scope on the input.

nventuro replied:
Same

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18586/commits/9d89862ba11f0be537d39b8afa4669ca5f5d1514

nventuro commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -637,7 +637,7 @@ export class PXEOracleInterface implements ExecutionDataProvider {
     nullifier: Fr,
     txHash: TxHash,
-    recipient: AztecAddress,
+    scope: AztecAddress,
```
```suggestion
    recipient: AztecAddress,
```

From the point of view of the contract, this is still the recipient. It is PXE that then chooses to store the note under the recipient's scope.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/pxe_oracle_interface.ts:
```diff
@@ -691,10 +691,9 @@ export class PXEOracleInterface implements ExecutionDataProvider {
     );
 
-    await this.noteDataProvider.addNotes([noteDao], recipient);
+    await this.noteDataProvider.addNotes([noteDao], scope);
```
```suggestion
    // The note was found by `recipient`, so we use that as the scope when storing the note.
    await this.noteDataProvider.addNotes([noteDao], recipient);
```

nventuro commented on yarn-project/pxe/src/storage/note_data_provider/note_data_provider.ts:
```diff
@@ -194,10 +194,10 @@ export class NoteDataProvider {
 
         if (scopes.length === 0) {
-          scopes = [dao.recipient.toString()];
+          throw new Error(`No scopes found for nullified note with index ${noteIndex}`);
```
```suggestion
          // We should never run into this error because notes always have a scope assigned to them - either
          // on initial insertion via `addNotes` or when removing their nullifiers.
          throw new Error(`No scopes found for nullified note with index ${noteIndex}`);
```

nventuro reviewed (approved):
Lovely. Please add a migration note mentioning the change to `getNotes`.

benesjan committed:
refactor: dropping recipient from `NoteDao` and `UniqueNote`

Fixes F-92

The issue description describes the motivation for this well:

> NoteDao contains a recipient value. This is no longer used as there is no filtering by recipient. It is also incorrect as notes don't technically have a recipient, they're just data. It is the note message that has a recipient.
>
> As recipient gets removed from the object, we'll no longer be able to use it to determine the note's scope. We should therefore ask the caller to specify the scope, which they can just do with the recipient field they used when creating the NoteDao (the message recipient).

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
